templates/agent/client/components/chat-history/AgentChangeHistoryItems.tsx

```diff
@@ -0,0 +1,179 @@
+		})
+	}, [items, editor])
+
+	const acceptance = useMemo<AgentChangeHistoryItem['acceptance']>(
+		() => items[0].acceptance,
+		[items]
+	)
+	const acceptanceConsensus = useMemo(
+		() => items.every((item) => item.acceptance === acceptance),
+		[items, acceptance]
+	)
+
+	return (
+		<div className="agent-change-message">
+			<div className="agent-change-message-actions">
+				{!acceptanceConsensus ? (
+					<span className="agent-change-message-acceptance-notice">Error</span>
+				) : (
+					<>
+						<button onClick={handleReject} disabled={acceptance === 'rejected'}>
+							<p>{acceptance === 'rejected' ? 'Rejected' : 'Reject'}</p>
+						</button>
+						<button onClick={handleAccept} disabled={acceptance === 'accepted'}>
+							<p>{acceptance === 'accepted' ? 'Accepted' : 'Accept'}</p>
+						</button>
+					</>
+				)}
+			</div>
+			<TldrawViewer shapes={diffShapes} components={{ ShapeWrapper: DiffShapeWrapper }} />
+		</div>
+	)
+}
+
+function getDiffShapesFromDiff({ diff }: { diff: RecordsDiff<TLRecord> }): TLShape[] {
+	const diffShapes: TLShape[] = []
+
+	for (const key in diff.removed) {
+		const id = key as TLShapeId
+		const prevShape = diff.removed[id]
+		if (prevShape.typeName !== 'shape') continue
+		const shape = {
+			...prevShape,
+			props: { ...prevShape.props },
+			meta: { ...prevShape.meta, changeType: 'delete' },
+		}
+
+		if ('dash' in shape.props) {
+			shape.props.dash = 'solid'
+		}
+
+		diffShapes.push(shape)
+	}
+
+	for (const key in diff.updated) {
+		const id = key as TLShapeId
+
+		const prevBefore = diff.updated[id][0]
+		const prevAfter = diff.updated[id][1]
+		if (prevBefore.typeName !== 'shape' || prevAfter.typeName !== 'shape') continue
+
+		const before = {
+			...prevBefore,
+			id: (id + '-before') as TLShapeId,
+			opacity: prevAfter.opacity / 2,
+			props: { ...prevBefore.props },
+			meta: { ...prevBefore.meta, changeType: 'update-before' },
+		}
+		const after = {
+			...prevAfter,
+			props: { ...prevAfter.props },
+			meta: { ...prevAfter.meta, changeType: 'update-after' },
+		}
+
+		if ('dash' in before.props) {
+			before.props.dash = 'dashed'
+		}
+		if ('fill' in before.props) {
+			before.props.fill = 'none'
+		}
+		if ('dash' in after.props) {
+			after.props.dash = 'solid'
+		}
+
+		diffShapes.push(before)
+		diffShapes.push(after)
+	}
+
+	for (const key in diff.added) {
+		const id = key as TLShapeId
+		const prevShape = diff.added[id]
+		if (prevShape.typeName !== 'shape') continue
+		const shape = {
+			...prevShape,
+			props: { ...prevShape.props },
+			meta: { ...prevShape.meta, changeType: 'create' },
+		}
+		if ('dash' in shape.props) {
+			shape.props.dash = 'solid'
+		}
+		diffShapes.push(shape)
+	}
+
+	return diffShapes
+}
+
+const DiffShapeWrapper = forwardRef(function DiffShapeWrapper(
+	{ children, shape, isBackground }: TLShapeWrapperProps,
+	ref: React.Ref<HTMLDivElement>
+) {
+	const changeType = useValue('change type', () => shape.meta.changeType, [shape])
```

Comment:
`shape` is a plain old js object, this will rerender with the shape anyway, so `useValue` is a no-op here